python/ray/serve/_private/http_proxy.py

```diff
@@ -199,12 +199,20 @@ def match_route(
         return None
 
 
-class HTTPProxy:
-    """This class is meant to be instantiated and run by an ASGI HTTP server.
-
-    >>> import uvicorn
-    >>> controller_name = ... # doctest: +SKIP
-    >>> uvicorn.run(HTTPProxy(controller_name)) # doctest: +SKIP
+class GenericProxy:
+    """This class is served as the base class for different types of proxies.
+    It contains all the common setup and methods required for running a proxy.
+
+    The proxy subclass need to implement the following methods:
+      - `proxy_name()`
+      - `_not_found()`
+      - `_draining_response()`
+      - `_timeout_response()`
+      - `_routes_response()`
+      - `_health_response()`
+      - `send_request_to_replica_unary()`
+      - `setup_request_context_and_handle()`
+      - `send_request_to_replica_streaming()`
```

Comment:
why do these need to be different across the impls?

Reply:
Bc how gRPC and ASGI handling request and response are very different. `_assign_request_with_timeout()` is still reused, but unwrapping receive object and construct response al need to be implemented differently in `send_request_to_replica_streaming()`